### ISSUE TITLE
feat: Ongoing campaign complex rules

### ISSUE THREAD

**Issue by muhsin-k** [chatwoot/chatwoot#3192]
@muhsin-k commented on [Sat Aug 14 2021](https://github.com/chatwoot/product/issues/158)

## **What is an ongoing campaign ?**

The ability for customer send outbound messages to their contacts which would trigger more conversations.

It is essentially a welcome message configured to pop up based on various conditions. Now are primarily focusing on triggering messages on simple page rules ( visitor visits page x / visitor spends x time of page x ). Now it supports on website live chat only

![campaign](https://user-images.githubusercontent.com/12408980/136735221-6e35f867-e2d2-4724-ab6e-96b9e1fd4c18.png)

[_https://www.notion.so/chatwoot/Proactive-Messaging-v1-a90f591623c04b9ab9e968060b28a41c_](https://www.notion.so/Proactive-Messaging-v1-a90f591623c04b9ab9e968060b28a41c)

## **Problem Statement**

Currently, the campaigns are working based on absolute URLs and the time they are spending on the website. But most real-world use cases consist of dynamic URLs.

For an example of an E-commerce website, the product ID for each product gives a different URL for the same page:

`https://ecart/products/23931/details
https://ecart/products/*/details`

At the moment it is not possible to run a campaign in a particular product since the ongoing campaign support 2 rules only

- The matching page URL
- Time on page

Related issues we are going to solve,

[Feature: Trigger Outgoing Messages based on events · Issue #657 · chatwoot/chatwoot](https://github.com/chatwoot/chatwoot/issues/657)

[Provide wildcard URL's for campaigns · Issue #3129 · chatwoot/chatwoot](https://github.com/chatwoot/chatwoot/issues/3129)

[feat: Allow wildcard URL's for campaigns · Issue #2739 · chatwoot/chatwoot](https://github.com/chatwoot/chatwoot/issues/2739)

### **Solution**

- Provide wildcard matching of URLs. Such a solution will reduce the effort required by the campaign creator when running it on multiple URLs.
- Add the ability to pre-configure ongoing campaigns in chatwoot for specific events and more logical conditions.

The conditions which we are going to support, 

1. Page URL —> is equal to/ start with/ contain/ doesn't contain → value (String)
2. Time on the current page → is equal to —> "n" seconds (Number)
3. Number of visits → is equal to/ greater than → value  (Number) (It will cover user first visit and the user visiting nth time)

## **Things to do**

- [ ] Add the option to add more logical conditions in add/edit campaign modal
- [ ] Update campaign emit events based on updated logical conditions
- [ ] Migrate existing campaigns to newly created schema

## **Technical changes**

- No change is needed in the backend since we already have the `trigger_rules` object in the campaign model.

**Trigger rules** 

attribute_key —> page_url/time_on_page/no. of visits

type —> string/ number

filter_operator —> equal_to, not_equal_to, contains, does_not_contain, less_than, greater_than

value —> facebook

join_type → AND, OR

Ex: 

page_url contains  pricing AND time_on_page equals 10 seconds OR number_of_visits equals 20 

**Rules structure**

```jsx
[
  {
    "attribute_key": "page_url",
    "type": "String",
    "filter_operator": "contains",
    "values": "chatwoot",
    "query_operator": "AND"
  },
  {
    "attribute_key": "time_on_page",
    "type": "Number",
    "filter_operator": "equal_to",
    "value": 5,
    "query_operator": "OR"
  },
  {
    "attribute_key": "number_of_visits",
    "type": "Number",
    "filter_operator": "equal_to",
    "values": 10,
    "query_operator":  null
  }
]
```

**Supported filter operators**

Attribute | Operators
--- | --- | 
page_url | equal_to, not_equal_to, contains, does_not_contains
time_on_page | equal_to
number_of_visits | equal_to, less_than, greater_than

**Validations**

- Minimum one rule is required

## **Additional questions**

- Think about moving add/edit campaign to a separate page

**Comment by tejaswinichile** [comment]
MVP: We can add multiple checks while launching the campaign, right now we just check the time attributes and the URL.

**Comment by brundles** [comment]
Here, could you also consider allowing for a RegEx @muhsin-k for the Page URL. That would be a simple way to give advanced users MANY more options.

**Comment by muhsin-k** [comment]
We have added the support wild card URL. It is available in the latest release (v2.12.o). More details can be found here https://www.chatwoot.com/docs/product/features/campaigns/wild-card-url-in-campaigns
